dmn-engine/src/main/scala/org/camunda/dmn/evaluation/DecisionTableEvaluator.scala

```diff
@@ -45,42 +48,41 @@ class DecisionTableEvaluator(
 
   def eval(decisionTable: ParsedDecisionTable,
            context: EvalContext): Either[Failure, Val] = {
-    implicit val ctx = context
-
-    evalInputExpressions(decisionTable.inputs).right.flatMap { inputValues =>
-      checkRules(decisionTable.rules, inputValues).right
-        .map(_.flatten)
-        .flatMap { matchedRules =>
-          matchedRules match {
-            case Nil => {
-              applyDefaultOutputEntries(decisionTable.outputs).right.map {
-                result =>
-                  audit(decisionTable, inputValues, Nil, result)
-              }
-            }
-            case rules if (decisionTable.hitPolicy == HitPolicy.FIRST) => {
+    implicit val ctx: EvalContext = context
+
+    evalInputExpressions(decisionTable.inputs) match {
+      case Left(failure) => // bad input values
+        evalAndAudit(decisionTable, Nil, Nil, Left(failure))
+      case Right(inputValues) =>
+        (checkRules(decisionTable.rules, inputValues) match {
+          case Left(failure) => // bad rules
+            evalAndAudit(decisionTable, inputValues, Nil, Left(failure))
+            Left(failure)
+          case right =>
+            right
+        }).map(_.flatten)
+          .flatMap {
+            case Nil =>
+              evalAndAudit(decisionTable,
+                           inputValues,
+                           Nil,
+                           applyDefaultOutputEntries(decisionTable.outputs))
+            case rules if (decisionTable.hitPolicy == HitPolicy.FIRST) =>
               val firstMatchedRule = List(rules.head)
-
-              evalOutputValues(firstMatchedRule).right.flatMap { values =>
-                applyHitPolicy(decisionTable, values.map(_.toMap)).right.map {
-                  result =>
-                    audit(decisionTable,
-                          inputValues,
-                          firstMatchedRule.zip(values),
-                          result)
-                }
+              evalOutputValues(firstMatchedRule).flatMap { values =>
+                evalAndAudit(decisionTable,
+                             inputValues,
+                             firstMatchedRule.zip(values),
+                             applyHitPolicy(decisionTable, values.map(_.toMap)))
               }
-            }
-            case rules => {
-              evalOutputValues(rules).right.flatMap { values =>
-                applyHitPolicy(decisionTable, values.map(_.toMap)).right.map {
-                  result =>
-                    audit(decisionTable, inputValues, rules.zip(values), result)
-                }
+            case rules =>
+              evalOutputValues(rules).flatMap { values =>
+                evalAndAudit(decisionTable,
+                             inputValues,
+                             rules.zip(values),
+                             applyHitPolicy(decisionTable, values.map(_.toMap)))
               }
```

Comment:
We could remove the pattern matching by introducing a context object. It collects all intermediate results and is used to build the audit result.

It is not the nicest design. But I think it improves visibility by separating the eval from the audit part. 

```
    class AuditContext(
                        var inputValues: List[Val],
                        var matchedRules: List[(ParsedRule, Iterable[(String, Val)])]
                      )

    val auditContext = new AuditContext(Nil, Nil)

    val result = evalInputExpressions(decisionTable.inputs).flatMap { inputValues =>
      auditContext.inputValues = inputValues

      checkRules(decisionTable.rules, inputValues)
        .map(_.flatten).flatMap {
        case Nil => {
          auditContext.matchedRules = Nil

          applyDefaultOutputEntries(decisionTable.outputs)
        }
        case matchedRules => {
          val rules = if (decisionTable.hitPolicy == HitPolicy.FIRST) {
            List(matchedRules.head)
          } else {
            matchedRules
          }

          evalOutputValues(rules).flatMap { values =>
            auditContext.matchedRules = rules.zip(values)

            applyHitPolicy(decisionTable, values.map(_.toMap))
          }
        }
      }

    }
    
    audit(decisionTable, auditContext.inputValues, auditContext.matchedRules, result)
    result
```

Reply:
I agree looks better now (despite the vars;))